### ISSUE TITLE
Can't change the capture screen shortcut on MacOS

### ISSUE THREAD

**Issue by mkrou** [flameshot-org/flameshot#1259]
**Flameshot version**
265e42c7b08e2296a55061b7bde61e0d56dbac3c

**System Information**
macOS Big Sur 11.1

**Problem**
It's impossible to change the capture screen shortcut in the gui configuration.

![image](https://user-images.githubusercontent.com/33038104/105812273-496ddf00-5fc7-11eb-81b8-46897cb733b8.png)

How can I change the shortcut? cmd+shift+alt+4 is super hard to use

**Comment by borgmanJeremy** [comment]
@thepurple any ideas on this?

**Comment by DirkSchneemann-TomTom** [comment]
Same thing in MacOS Catalina 10.15.7, if that is of any help in debugging.

**Comment by thepurple** [comment]
Sorry for a long response, I missed this ticket.
It's not an issue, it's a feature request.
Currently it's hard-coded and is not configurable. 
Will add this feature it later.

**Comment by mkrou** [comment]
@thepurple could you please show where is it hardcoded? I can't find it.

**Comment by thepurple** [comment]
@mkrou 
https://github.com/flameshot-org/flameshot/blob/b5ca48dde0102a6a6256c65762dffe582d05996e/src/core/controller.cpp#L101

**Comment by borgmanJeremy** [comment]
Similar request for windows #1341

**Comment by isheraz** [comment]
Will there be an update soon?

**Comment by jaredyam** [comment]
The current shortcut for taking screenshots (⌘ + ⌥ + shift + 4) is really hard to use. Maybe the default shortcut used in the Xnip (⌘ + shift + X) will be a better solution.

**Comment by isheraz** [comment]
> The current shortcut for taking screenshots (⌘ + ⌥ + shift + 4) is really hard to use. Maybe the default shortcut used in the Xnip (⌘ + shift + X) will be a better solution.

Yes a smaller one would be nicer

**Comment by danielgroen** [comment]
`⌘ + shift + 4`

**Comment by mmahmoudian** [comment]
@isheraz there will be an update soon, but I'm not sure if this issue is addressed in it. Perhaps @thepurple can comment as he knows better that me what is the current state.

**Comment by wittd19** [comment]
FYI - On Mac, I was able to map the PrintScreen button using my logitech keyboard software to the default key-mapping.

**Comment by ctaetcsh** [comment]
Bumping this issue cause its been like a month since this has last had any activity and this issue is still not fixed. If it's coming as a feature, why can't you just unlock it now? Was this fixed in a newer version and mine is just out of date or?

tldr: my fingers f**king hurt pls fix

**Comment by borgmanJeremy** [comment]
We can't unlock it now because the work is not done. We have busy lives and work on this for free as our time allows. 

If you want it done now, do it yourself and submit a PR. If you want guidance on implementing it hit us up on the slack.

**Comment by ctaetcsh** [comment]
okie dokie

**Comment by mmahmoudian** [comment]
@ctaetcsh Flameshot is an OpenSource software and like most FLOSS softwares out there, it is done by volunteer work. I hope you understand why development is not as rapid as proprietary software.

Now back to your question, the latest state of the project can be easily compiled from source. You can find proper instructions here:

https://flameshot.org/guide/installation/compilation/#osx

**Comment by ctaetcsh** [comment]
@mmahmoudian I understand that this is done via volunteer work and that progress is slower then with software that has people paid to work on it. The attitude of my original comment is mostly because it seems like a weird thing to take a long time to fix. Why are those last 5 shortcuts hard coded? Is it a bodge from an other issue or just an oversight? 

I have already started to see if I could just fix it for myself by changing the hard coded default to something that would work for me, and I appreciate you giving me pointers on where to begin since I've never dabbled with C++ stuff before.

**Comment by ctaetcsh** [comment]
Okay I was able to change the hard coded default to F13 and it's working great!
Looking at configshortcuts.cpp, line 89 to 91 includes the following comment:
```
    // TODO - Linux doesn't support global shortcuts for (XServer and Wayland),
    // possibly it will be solved in the QHotKey library later. So it is
    // disabled for now.
```
I assume it was just disabled across everything until the issue could be looked into further. If that was so, couldn't a temporary solution be to just disable it based on the OS?

**Comment by mmahmoudian** [comment]
@ctaetcsh  I'm not sure if I understood your point, but the code already tries to handle different operating systems differently:
![Screenshot of the code with annotations](https://user-images.githubusercontent.com/390889/121847379-2a12ec80-ccf1-11eb-9356-e62a176b14ce.jpeg)

And in Linux, as far as I know, there is truly no possibility of defining global shortcuts because the shortcuts are managed by the window manager (much much more customisable compared to MacOS or Windows). Flameshot was initially designed for Linux and we recent started supporting Windows and Mac.

**Comment by ctaetcsh** [comment]
If it can change based on OS, why is it disabled (i'm making an assumption here) on any OS and not just Linux where this issue exists? If that is the solution, that doesn't seem like a particularly difficult thing to implement and thus cross off a 5 month old issue.

**Comment by mmahmoudian** [comment]
This discussion is definitely off-topic, but FYI, in Linux the shortcuts for starting software are not handled by the software itself but rather the [window manager](https://en.m.wikipedia.org/wiki/Window_manager) (which is something that can easily changed and customized in Linux but it is impossible to change or modify freely in Windows and OSX). Additionally, Flameshot uses QT library to communicate with the window manager, and because [there are many window managers in Linux](https://linux.fandom.com/wiki/Window_managers), as far as I know, there is no unified way to implement global shortcuts and therefore QT doesn't have a ready-made functionality to set it.

Believe it or not, we have never received any of such issues from Linux users and everything works perfectly. All such filed issues come from users with limited and restricted operating systems (i.e Windows and MacOS).

Regardless, as you saw this is relatively an easy thing to change and compile.

**Comment by kossa** [comment]
@thepurple any update about this, I found this cool tool on mac, everything well except for the shortcut, really it should be customizable, it's the most important, you can hard other shortcut 😁

**Comment by thepurple** [comment]
@kossa no updates yet, sorry. I'll try to find some time for this, but without ETA for now :)

**Comment by kossa** [comment]
Excellent thank you, i'll still work with it 👌

**Comment by ctaetcsh** [comment]
@kossa Meant to reply to this earlier, but as I showed above it's easy to change the binding and recompile it with that new hard coded binding. This is what I have been doing and it has worked flawlessly.

**Comment by kossa** [comment]
Thank you @ctaetcsh I dont have GCC environment and dont want to install it, all screenshot apps offer the possibility to customize the shortcut, and Flameshot is a great one I used so it should have this feature 👍

**Comment by OrkunTokdemir** [comment]
it will make it configurable.

**Comment by Markkop** [comment]
Hey, there! What if it's possible to trigger the capture action by biding a shortcut in this Keyboard Config?
<img width="632" alt="image" src="https://user-images.githubusercontent.com/16388408/130492170-e6be1abd-3c0a-4a56-8ac3-717fe510dd5b.png">

**Comment by kossa** [comment]
It does not work for me

**Comment by troy-lamerton** [comment]
This GitHub issue has many participants so please keep off-topic discussions to a minimum.

The collaborators have already pointed out where to change the hard-coded value.

I would propose this:

- Change [hard-coded shortcut](https://github.com/flameshot-org/flameshot/blob/v0.10.1/src/core/controller.cpp#L95-L96) to something easy
- `⌘ + shift + X` is recommended in comment https://github.com/flameshot-org/flameshot/issues/1259#issuecomment-808661441

This one-liner change would improve the experience a lot for us Mac users. No need to check it compiles if you only change the string.

If I had time I would make a PR myself. Later we could consider an option to change the keyboard shortcut.

**Comment by ctaetcsh** [comment]
Seeing these updates, it's hard not to get somewhat frustrated because as far as I can tell this is what we've learned:
* The key combination can be set to basically anything and have it work fine
* The code and UI work to change key shortcuts is there
I can't be the only one who feels like the solution here is fairly straight forward. Telling people to "go do the work and open a PR" isn't helpful, not everyone is an expert in C and Qt. More than that, there hasn't been a clear explanation as to why these specific shortcuts are disabled.

Now, given that all of what I just is considered "off-topic", I ask for one of two things:
* An explanation (at your leisure) as to why things are this way
or
* A commitment to have somebody look into this and fix the issue.

Again, this is a request, NOT a demand. I understand that we are not entitled to either, but the request to fix this issue is made with (what I hope) are good intentions in wanting to help improve this software. We are not your enemy; we have an interest in this software becoming better too.

I appreciate your time. Thank you.

**Comment by borgmanJeremy** [comment]
There's no contention on changing this to be configurable. I don't have easy access to a Mac to do testing on so I am prioritizing other issues with my limited free time.

**Comment by ctaetcsh** [comment]
Thank you.

**Comment by truefan4ever** [comment]
It would be cool to change shortcut, waiting for this feature.

**Comment by deseven** [comment]
For those who use [Karabiner-Elements](https://karabiner-elements.pqrs.org), here's the config to map ⇧⌘T to ⇧⌘⌥4 (you can change ⇧⌘T to anything you want):

```
{
  "title": "Flameshot global hotkey workaround",
  "rules": [
    {
      "description": "Map ⇧⌘T to ⇧⌘⌥4",
      "manipulators": [
        {
          "type": "basic",
          "from": {
            "key_code": "t",
            "modifiers": {
              "mandatory": [
                "left_shift",
                "left_command"
              ]
            }
          },
          "to": [
            {
              "key_code": "4",
              "modifiers": [
                "left_command",
                "left_shift",
                "left_option"
              ]
            }
          ]
        }
      ]
    }
  ]
}
```

Put it in `~/.config/karabiner/assets/complex_modifications/flameshot.json` and enable in the "Complex modifications" tab of Karabiner-Elements Preferences.

**Comment by borgmanJeremy** [comment]
This is fixed on the master branch now and will be part of version 11.

**Comment by up-to-you** [comment]
It would be nice to install v11 this year !

Do i have a chance to ask about approximate release date ?

P.s. Definitely would be a contributor if it was written in Rust (no experience with segfault-free C++ 😁 )

**Comment by borgmanJeremy** [comment]
I'm a fellow rust enthusiast 😁 if the FFI between rust and c++ was a little cleaner I would consider writing the new upload feature in rust 

I was planning to release the beta last weekend but some personal issues came up. Hopefully this week/weekend I can get it out. We already feature froze so only merging bug fixes until the release is out.

**Comment by up-to-you** [comment]
https://github.com/google/autocxx is an insane tool as a C++/Rust glue.

P.s. just to increase chance for Rust integration one more percent 😁

**Comment by EhsanQA62** [comment]
Keyboard > Keyboard Shortcuts > App Shortcuts
![Uploading image.png…]()

**Comment by Enstrayed** [comment]
@robertasDS This issue is long since resolved, if you're encountering a similar issue on Windows you'll likely need to find/open a new issue for Windows specifically. However, if you are on Windows, you may consider other software with similar functionality to Flameshot where keybind remapping is thoroughly supported:
* [ShareX](https://getsharex.com/)
* [ksnip](https://github.com/ksnip/ksnip)

**Comment by tolache** [comment]
For anyone still looking to bind it to ⇧⌘4 (Command + Shift + 4), you don't actually need Karabiner Elements or any third party software for this. You can just to unbind the standard macOS action for this shortcut in System Settings > Keyboard > Keyboard shortcuts. Then you will be able to assign this key combination to screen capture in Flameshot settings.
